### PR TITLE
Remove Iterable<Integer>

### PR DESCRIPTION
**Pull request by Maithem**
Implementing Iterable<T> allows applications to use  enhanced for
loops, syntactic sugar, at the cost of performance (caused by
auto-boxing). Since RoaringBitmap already implements forEach with
IntConsumer, this patch removes the Iterable interface implementation
from RoaringBitmap. This should encourage applications to use more
efficient APIs and existing applications that use the enhanced for
loop can easily convert the code to loop the iterator directly.

### CODE REVIEW

**Comment by Maithem** [comment]
Addresses #459

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java
Comment (blacelle) [inline RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java:2057]: I feel the method should be dropped alltogether, or at least marked `@Deprecated`

File: RoaringBitmap/src/main/java/org/roaringbitmap/buffer/ImmutableRoaringBitmap.java
Comment (blacelle) [inline RoaringBitmap/src/main/java/org/roaringbitmap/buffer/ImmutableRoaringBitmap.java:1460]: I feel the method should be dropped alltogether, or at least marked `@Deprecated`

File: RoaringBitmap/src/test/java/org/roaringbitmap/TestRoaringBitmap.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/TestRoaringBitmap.java:2987]: We should demonstrate the usage without autoboxing, probably through .forEach(int -> XXX)

**Review by blacelle — CHANGES_REQUESTED** [review]
I feel we should demove or deprecated the various iterator() method, and demontrate in our own code how to migrate.

**Comment by richardstartin** [comment]
I agree, just deleting the code is probably the least surprising thing to do when someone upgrades, and keeping the method on the base class is still binary incompatible anyway. I don't know if users should be directed towards `forEach` though - `BatchIterator` is the fastest API available, just a little awkward to use in modern Java.

One thing that concerns me about all of these changes is - despite making total sense _now_ - is recalling the mess I encountered in Hadoop projects 5 or 6 years ago because of Guava's historical deletion-over-deprecation policy. This library is used in a lot of places and in the kind of projects prone to transitive dependency conflicts. Along with these rationalisations we should probably rename the package before releasing.

**Comment by blacelle** [comment]
I love the Hadoop+Guava argument ;). I observed the same kind of issue with heterogenous Spark clusters, in which RoaringBitmap is definitely used. Keeping the interface, but marking it as deprecated is very fine with me.

People with strong performance considerations shall be burdened by calling deprecated methods.

**Comment by lemire** [comment]
So we do not delete but deprecate, correct?

**Comment by richardstartin** [comment]
@lemire 

> So we do not delete but deprecate, correct?

If you change the package name you can do whatever you want. Deprecating is hard to do (impossible?) with the `LongConsumer`/`IntConsumer` change because there's a clash between a method override/hide which needs to be resolved with type information. So either we stay with the structurally identical/nominally distinct `LongConsumer`/`IntConsumer` types and make none of these changes, or we make these changes and change the package name.

**Comment by lemire** [comment]
Do you have a specific proposal regarding how we should rename the package?

**Comment by richardstartin** [comment]
`org.roaringbitmap1` should do it. I think every library should do this sort of thing - the tendency is to break (Guava) or accrue backward compatible garbage. Again, this rationalisation can't be done backward compatibly, so the tendency would be to just not do it at all even though it's a big improvement...

**Comment by lemire** [comment]
Ok. What do people think of `org.roaringbitmap1`? 

@blacelle ?

**Comment by blacelle** [comment]
Not a fan of `org.roaringbitmap1`.

Here is a list of packages which version in them:

    com.sun.xml.bind.v2.model.annotation
    com.google.iam.v1
    zipkin2.v1.V1Span
    org.apache.commons.lang3

My preference would be: `org.roaringbitmap` -> `org.roaringbitmap.v2` to maintain the existing parent packege, and enable future versions.

**Comment by lemire** [comment]
@richardstartin  What do you think of `org.roaringbitmap.v2`?

**Comment by richardstartin** [comment]
Sounds good, same principle

**Comment by lemire** [comment]
I have changed the base for this PR to https://github.com/RoaringBitmap/RoaringBitmap/tree/1.0-SNAPSHOT

I suggest that once @blacelle  approves it, we merge it to 1.0-SNAPSHOT then proceed, separately, with a package renaming.

**Comment by blacelle** [comment]
I would approve once iterator implementation is also removed.

**Comment by Maithem** [comment]
@blacelle the problem with removing the Iterator is that variables from outside the lambda need to be either final/array, or some object. For example, this would produce a compilation error:

```java 

int x = 0;
rb.forEach(value -> {
   x += value;
});

```

**Comment by blacelle** [comment]
I agree. Still, this is the point of this evolution. This is why I proposed introducing an IntStream. One may decide the sum through an AtomicInteger. For a small unit-test, one may go with .toArray()

**Comment by blacelle** [comment]
(Just in case, IntStream is a candidate for a future PR, not for this one).

**Comment by richardstartin** [comment]
One of the oldest tricks in the book for reductions within closures is

```java
int[] x = {0};
rb.forEach(value -> {
   x[0] += value;
});
```

You're always better off doing this than boxing, I don't even find it particularly dirty. Any serious use case of this library would avoid the `Iterable<Integer>` access method at all costs, and some users eschew convenience entirely - e.g. Druid picked up `BatchIterator` for its performance in their vectorised query engine, despite its extremely awkward API. 

In my opinion, if your top priority is being able to write

```java
int x = 0;
for (Integer i : set) {
  x += value;
}
```

then `HashSet` is probably what you're looking for...

**Comment by richardstartin** [comment]
@blacelle you can already stream the contents of the bitmap https://github.com/RoaringBitmap/RoaringBitmap/commit/b4f75f6304afeb4d5240e17e4299a95e76ac4d08

**Comment by blacelle** [comment]
> @blacelle you can already stream the contents of the bitmap b4f75f6

BANG @richardstartin , I forgot this.

@Maithem Go for `rb.stream().sum();`

**Comment by Maithem** [comment]
@blacelle  @richardstartin please take a look.

File: RoaringBitmap/src/test/java/org/roaringbitmap/TestIteratorMemory.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/TestIteratorMemory.java:110]: This is not equivalent to the original code in case of overflow. For safety, you should have some thing like bitmap_a.stream().mapToLong(i -> i).sum();

https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8073933

File: RoaringBitmap/src/test/java/org/roaringbitmap/TestIterators.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/TestIterators.java:88]: This makes me think one could migrate from `bitmap.iterator() `to `bitmap.stream().iterator()`

I suggest keeping the original code, and switching to this alternative way to get an `Iterable<Integer>`. Sorry for not having throught about it earlier.

File: RoaringBitmap/src/test/java/org/roaringbitmap/TestSerialization.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/TestSerialization.java:211]: Given we are in a test, I would have gone for a double iteration, andf go through:

`bitmap_a.stream().sum()`; (by the way, here, we do not care with int overflow).
`bitmap_a.stream().count()`.

File: RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestImmutableRoaringBitmap.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestImmutableRoaringBitmap.java:1510]: I would keep the variable name `next `instead of `x`, for the sake of smaller diff.

File: RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestIterators.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestIterators.java:260]: We do not care about overflow here, OK

File: RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestMemoryMapping.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestMemoryMapping.java:556]: identation KO

File: RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestRoaringBitmap.java
Comment (blacelle) [inline RoaringBitmap/src/test/java/org/roaringbitmap/buffer/TestRoaringBitmap.java:2909]: This may be kept as a test over bitmap.stream().iterator()

File: fuzz-tests/src/test/java/org/roaringbitmap/BufferFuzzer.java
Comment (blacelle) [inline fuzz-tests/src/test/java/org/roaringbitmap/BufferFuzzer.java:10]: Unclear need for this import

File: fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java
Comment (blacelle) [inline fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java:537]: Indentation KO

**Review by blacelle — APPROVED** [review]
Good progress, but still a few things.

@lemire @richardstartin It appears an Iterabler<Integer> is still easily available through bitmap.stream().iterator()

**Review by blacelle — CHANGES_REQUESTED** [review]
I did not mean 'approve' but 'request changes'

**Comment by blacelle** [comment]
@Maithem Would you mind also purging 64bits implementations? Thanks

**Comment by lemire** [comment]
Ping @Maithem  : This PR still has requested changes... Can you check?

**Comment by Maithem** [comment]
@lemire been busy, will take a look tonight.

**Pull request closed without merging**
